---
title: "data_analysis_mito"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## load libraries and read in necessary files
```{r, include=F}
library(tidyverse)
library(cowplot)
library(ape)
library(pegas)
library(RColorBrewer)
library(knitr)
library(insect)
source("/workdir/genomic-data-analysis/scripts/mtgenome_functions.R")

sample_table_mt <- read_tsv("/fs/cbsubscb16/storage/rkc/sample_lists/sample_table_mito.tsv")
```

## Count alleles and make vcf

```{bash}
## Get allele count
nohup /programs/angsd-0.940/angsd -bam /fs/cbsubscb16/storage/rkc/sample_lists/bam_list_mito.txt -doCounts 1 -minQ 20 -dumpCounts 4 -doBcf 1 -gl 1 -dopost 1 -domajorminor 1 -domaf 1 -out /fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20 > /fs/cbsubscb16/storage/rkc/nohups/bam_list_mito_minq20.nohup &

mv /fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20.counts.gz /fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20.allele_counts.gz

## Get depth count
nohup /programs/angsd-0.940/angsd -bam /fs/cbsubscb16/storage/rkc/sample_lists/bam_list_mito.txt -doCounts 1 -minQ 20 -dumpCounts 2 -out /fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20 > /fs/cbsubscb16/storage/rkc/nohups/bam_list_mito_minq20.nohup &

mv /fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20.counts.gz /fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20.depth_counts.gz
```

## Visualize depth statistics 

#### Depth per populations
```{r}
ind_label <- read_tsv("/fs/cbsubscb16/storage/rkc/sample_lists/sample_table_mito.tsv")$ABLG
pop_label <- read_tsv("/fs/cbsubscb16/storage/rkc/sample_lists/sample_table_mito.tsv")$population
loc_label <- read_tsv("/fs/cbsubscb16/storage/rkc/sample_lists/sample_table_mito.tsv")$Loc
depth_count <- read_tsv("/fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20.depth_counts.gz", col_select = 1:176) %>%
  t() %>%
  as_tibble() %>%
  bind_cols(ind=ind_label, population=pop_label, .) %>%
  gather(key = position, value="depth", 3:16627) %>%
  mutate(position=as.numeric(substring(position, 2)))
group_by(depth_count, ind, population) %>%
  summarise(average_depth=mean(depth)) %>%
  ggplot(mapping=aes(x=population, y=average_depth, group=population)) +
  geom_boxplot() +
  geom_jitter() +
  theme_cowplot() +
  coord_flip()
```

#### Depth by position 
```{r}
group_by(depth_count, population, position) %>%
  summarise(average_depth=mean(depth)) %>%
  ggplot(aes(x=position, y=average_depth, color=population)) +
  geom_line() +
  theme_cowplot()

```
- Depths look reasonable across samples and loci

## Get fasta file
```{r}

concensus_to_fasta_2 <- function(concensus, ind_label, out_path){
  ## This function is used to convert the output from convert_count_to_concensus() into a fasta file
  # concensus should be a data frame outputted by convert_count_to_concensus()
  # ind_label should be a vector with the same length as the number of rows in concensus, containing the individual labels for each sequence
  # out_path should be the output path of the fasta file
  for (i in seq_along(ind_label)){
    if (i==1) {
      write_lines(paste0(">", ind_label[i], "_", pop_label[i], "_", loc_label[i]), out_path, append = F)
    } else {
      write_lines(paste0(">", ind_label[i], "_", pop_label[i], "_", loc_label[i]), out_path, append = T)
    }
    write_lines(paste0(concensus[i,], collapse = ""), out_path, append = T)
  }
}

source("/workdir/genomic-data-analysis/scripts/mtgenome_functions.R")

## Produce fastas with increasingly stringent depth filters to test in downstream analyses
# minimum depth 1, minimum major allele frequency 0.75
concensus <- convert_count_to_concensus(x="/fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20.allele_counts.gz", min_depth=1, min_maf=0.75)
concensus_to_fasta_2(concensus, ind_label, "/fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20_mindepth1_minmaf75.fasta")

# minimum depth 2, minimum major allele frequency 0.75
concensus <- convert_count_to_concensus(x="/fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20.allele_counts.gz", min_depth=2, min_maf=0.75)
concensus_to_fasta_2(concensus, ind_label, "/fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20_mindepth2_minmaf75.fasta")

# minimum depth 4, minimum major allele frequency 0.75
concensus <- convert_count_to_concensus(x="/fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20.allele_counts.gz", min_depth=4, min_maf=0.75)
concensus_to_fasta_2(concensus, ind_label, "/fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20_mindepth4_minmaf75.fasta")
```

## Plot haplotype network (relaxed filter)
```{r}
fasta <- read.dna("/fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20_mindepth1_minmaf75.fasta", format="fasta")
fasta
SEAK_list <- c(5570:5580, 5593:5597, 5626:5630, 5649:5651, 5653, 5655:5658, 5717:5722)

hap <- haplotype(fasta)
hap_net <- haploNet(hap, d=dist.dna(hap, model = "N"))

ind_hap <- with(stack(setNames(attr(hap, "index"), rownames(hap))), table(hap=ind, individuals=rownames(fasta)[values])) %>%
  as_tibble() %>%
  mutate(population=substr(individuals,14,nchar(individuals)))

## number of positions retained
t(as.character(fasta)) %>% 
  as_tibble() %>%
  filter_all(all_vars(. != "n")) %>%
  dim() %>%
  .[1]
```

```{r}
## plotting 
pop_hap <- ind_hap %>%
  group_by(hap, population) %>%
  summarise(n=sum(n, na.rm = T)) %>%
  ungroup() %>%
  spread(population, n) %>%
  mutate_all(~replace(., is.na(.), 0)) %>%
  mutate(hap=as.roman(hap)) %>%
  arrange(hap) %>%
  select(-hap) %>%
  as.matrix()
pie_color <- brewer.pal(n = 5, name = "Paired")
set.seed(1)
plot(hap_net, size=sqrt(attr(hap_net, "freq"))*1.5, fast = F, scale.ratio = 1, pie=pop_hap, bg=pie_color, show.mutation=1, labels=F, threshold=c(0))
legend("topleft", c(colnames(pop_hap)), fill=pie_color, cex=0.8)
```

## Plot haplotype network (stringent filter)
```{r}
fasta <- read.dna("/fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20_mindepth4_minmaf75.fasta", format="fasta")
fasta

hap <- haplotype(fasta)
hap_net <- haploNet(hap, d=dist.dna(hap, model = "N"))

ind_hap <- with(stack(setNames(attr(hap, "index"), rownames(hap))), table(hap=ind, individuals=rownames(fasta)[values])) %>%
  as_tibble() %>%
  mutate(population=substr(individuals,14,nchar(individuals)))

## number of positions retained
t(as.character(fasta)) %>% 
  as_tibble() %>%
  filter_all(all_vars(. != "n")) %>%
  dim() %>%
  .[1]
```

```{r}
pop_hap <- ind_hap %>%
  group_by(hap, population) %>%
  summarise(n=sum(n, na.rm = T)) %>%
  ungroup() %>%
  spread(population, n) %>%
  mutate_all(~replace(., is.na(.), 0)) %>%
  mutate(hap=as.roman(hap)) %>%
  arrange(hap) %>%
  select(-hap) %>%
  as.matrix()
pie_color <- brewer.pal(n = 5, name = "Paired")
set.seed(1)
png("/fs/cbsubscb16/storage/rkc/figures/hapnet_mitogenome.png", height = 5, width = 8, units = "in", res = 200)
plot(hap_net, size=sqrt(attr(hap_net, "freq"))*1.5, fast = F, scale.ratio = 1, pie=pop_hap, bg=pie_color, show.mutation=1, labels=F, threshold=c(0))
legend("bottomleft", c(colnames(pop_hap)), fill=pie_color, cex=0.8)
dev.off()
```

## Haplotype network of Southeast AK

```{r}
fasta_SEAK <- read.dna("/fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20_mindepth1_minmaf75_SEAK.fasta", format="fasta")
fasta_SEAK
SEAK_list <- c(5570:5580, 5593:5597, 5626:5630, 5649:5651, 5653, 5655:5658, 5717:5722)

hap <- haplotype(fasta_SEAK)
hap_net <- haploNet(hap, d=dist.dna(hap, model = "N"))

ind_hap <- with(stack(setNames(attr(hap, "index"), rownames(hap))), table(hap=ind, individuals=rownames(fasta_SEAK)[values])) %>%
  as_tibble() %>%
  mutate(population=paste(substr(individuals,1,nchar(individuals)-4), (read_tsv("/fs/cbsubscb16/storage/rkc/sample_lists/sample_table_mito.tsv") %>% filter(Loc == "SEAK"))$population))

## number of positions retained
t(as.character(fasta_SEAK)) %>% 
  as_tibble() %>%
  filter_all(all_vars(. != "n")) %>%
  dim() %>%
  .[1]
```

```{r}
pop_hap <- ind_hap %>%
  group_by(hap, population) %>%
  summarise(n=sum(n, na.rm = T)) %>%
  ungroup() %>%
  spread(population, n) %>%
  mutate_all(~replace(., is.na(.), 0)) %>%
  mutate(hap=as.roman(hap)) %>%
  arrange(hap) %>%
  select(-hap) %>%
  as.matrix()
pie_color <- brewer.pal(n = 7, name = "Accent")
set.seed(1)
plot(hap_net, size=sqrt(attr(hap_net, "freq"))*1.5, fast = F, scale.ratio = 1, pie=pop_hap, bg=pie_color, show.mutation=1, labels=F, threshold=c(0))
legend("bottomleft", c(colnames(pop_hap)), fill=pie_color, cex=0.8)
```

## Try haplotype frequencies per population 

```{r}
## I think you have to use hapoFreq() from pegas package
## will help to reformat fasta file to include population and Loc info 
fasta_pops <- read.dna("/fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20_mindepth4_minmaf75_coi.fasta", format = "fasta")
hap_pops <- haplotype(fasta_pops)

freqs_loc <- haploFreq(fasta_pops, what = 3, haplo = hap_pops)
net_loc <- haploNet(hap_pops)
freq <- attr(net_loc, "freq")
plot(net_loc, size = freq*3, pie = freqs_loc, labels = F)

x <- as.data.frame(freqs_loc) %>% mutate(haplotype = seq(1:71))

x1 <- x %>% select(AI,haplotype) %>% filter(AI != 0)
x2 <- x %>% select(EastBering,haplotype) %>% filter(EastBering != 0)
x3 <- x %>% select(GOA,haplotype) %>% filter(GOA != 0)
x4 <- x %>% select(NorthBering,haplotype) %>% filter(NorthBering != 0)
x5 <- x %>% select(SEAK,haplotype) %>% filter(SEAK != 0)

op <- par(mfcol = c(3, 2))
ggplot(x1, aes(x="", y=AI, fill=as.character(haplotype))) +
  geom_bar(stat="identity", width=1, color = "black") +
  coord_polar("y", start=0) +
  labs(title = "Aleutian Islands") +
  cowplot::theme_cowplot() +
  theme(axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks = element_blank(), 
        axis.title = element_blank(), 
        plot.background = element_blank())
ggplot(x2, aes(x="", y=EastBering, fill=as.character(haplotype))) +
  geom_bar(stat="identity", width=1, color = "black") +
  coord_polar("y", start=0) +
  labs(title = "East Bering") +
  cowplot::theme_cowplot() +
  theme(axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks = element_blank(), 
        axis.title = element_blank(), 
        plot.background = element_blank())
ggplot(x3, aes(x="", y=GOA, fill=as.character(haplotype))) +
  geom_bar(stat="identity", width=1, color = "black") +
  coord_polar("y", start=0) +
  labs(title = "Gulf of Alaska") +
  cowplot::theme_cowplot() +
  theme(axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks = element_blank(), 
        axis.title = element_blank(), 
        plot.background = element_blank())
ggplot(x4, aes(x="", y=NorthBering, fill=as.character(haplotype))) +
  geom_bar(stat="identity", width=1, color = "black") +
  coord_polar("y", start=0) +
  labs(title="North Bering") +
  cowplot::theme_cowplot() +
  theme(axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks = element_blank(), 
        axis.title = element_blank(), 
        plot.background = element_blank())
ggplot(x5, aes(x="", y=SEAK, fill=as.character(haplotype))) +
  geom_bar(stat="identity", width=1, color = "black") +
  coord_polar("y", start=0) +
  labs(title = "Southeast Alaska") +
  cowplot::theme_cowplot() +
  theme(axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks = element_blank(), 
        axis.title = element_blank(), 
        plot.background = element_blank())
par(op)

x_long <- x %>% pivot_longer(!haplotype, names_to = "population", values_to = "count")
x_long$haplotype = factor(x_long$haplotype)
ggplot(x_long, aes(x="", y = count, fill = haplotype)) +
  geom_bar(stat="identity", width=1, color = "black") +
  coord_polar("y", start=0, clip = "off") +
  cowplot::theme_cowplot() +
  theme(axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks = element_blank(), 
        axis.title = element_blank(), 
        plot.background = element_blank()) +
  facet_grid(population ~ .)

#ggsave("/fs/cbsubscb16/storage/rkc/figures/hap_freqs_perPop.pdf", device = "pdf", width = 5, height = 7, units = "in")
```

## Haplonetwork with just COI

```{r}
fasta_coi <- read.dna("/fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20_mindepth4_minmaf75_coi.fasta", format="fasta")
fasta_coi

hap <- haplotype(fasta_coi)
hap_net <- haploNet(hap, d=dist.dna(hap, model = "N"))

ind_hap <- with(stack(setNames(attr(hap, "index"), rownames(hap))), table(hap=ind, individuals=rownames(fasta_coi)[values])) %>%
  as_tibble() %>%
  mutate(population=substr(individuals,14,nchar(individuals)))

## number of positions retained
t(as.character(fasta_coi)) %>% 
  as_tibble() %>%
  filter_all(all_vars(. != "n")) %>%
  dim() %>%
  .[1]
```

```{r}
pop_hap <- ind_hap %>%
  group_by(hap, population) %>%
  summarise(n=sum(n, na.rm = T)) %>%
  ungroup() %>%
  spread(population, n) %>%
  mutate_all(~replace(., is.na(.), 0)) %>%
  mutate(hap=as.roman(hap)) %>%
  arrange(hap) %>%
  dplyr::select(-hap) %>%
  as.matrix()
pie_color <- brewer.pal(n = 5, name = "Paired")
set.seed(1)
png("/fs/cbsubscb16/storage/rkc/figures/hapnet_COI.png", height = 5, width = 8, units = "in", res = 200)
plot(hap_net, size=sqrt(attr(hap_net, "freq"))*0.3, fast = F, scale.ratio = 1, pie=pop_hap, bg=pie_color, show.mutation=1, labels=F, threshold=c(0))
legend("bottomleft", c(colnames(pop_hap)), fill=pie_color, cex=0.8)
dev.off()
```
## include Grant and Zheng study individuals
```{r}
gz_fasta.coi <- read.dna("/fs/cbsubscb16/storage/rkc/angsd/bam_list_mito_minq20_mindepth4_minmaf75_grant_zheng_coi_align.fas", format = "fasta")

hap_coi <- haplotype(gz_fasta.coi)

net_coi <- haploNet(hap_coi)
freqs_coi <- haploFreq(gz_fasta.coi, what = 3, haplo = hap_coi)
freq <- attr(net_coi, "freq")
plot(net_coi, size = freq*3, pie = freqs_coi, labels = F)

x <- as.data.frame(freqs_coi) %>% mutate(hap = seq(1:144))

x1 <- x %>% dplyr::select(AI,hap) %>% filter(AI != 0)
x2 <- x %>% dplyr::select(EastBering,hap) %>% filter(EastBering != 0)
x3 <- x %>% dplyr::select(GOA,hap) %>% filter(GOA != 0)
x4 <- x %>% dplyr::select(NorthBering,hap) %>% filter(NorthBering != 0)
x5 <- x %>% dplyr::select(SEAK,hap) %>% filter(SEAK != 0)

ggplot(x, aes(x="", y = camtschaticus, fill=as.character(hap))) +
  geom_bar(stat="identity", width=1, color = "black") +
  coord_polar("y", start=0) +
  labs(title = "Aleutian Islands")
```

## Haplotype diversity

```{r}
## split pops in fasta DNAbin object
fasta_labs <- labels.DNAbin(fasta)
AI_index <- grep("AI", fasta_labs)
EBS_index <- grep("EastBering", fasta_labs)
GOA_index <- grep("GOA", fasta_labs)
NBS_index <- grep("NorthBering", fasta_labs)
SEAK_index <- grep("NorthBering", fasta_labs)

# get haplotype diversity per population
hap.div(fasta[AI_index,]) # 0.9916667
hap.div(fasta[EBS_index,]) # 0.9943101
hap.div(fasta[GOA_index,]) # 0.9970777
hap.div(fasta[NBS_index,]) # 1.00
hap.div(fasta[SEAK_index,]) # 1.00
```

