---
title: "Diversity genome scan"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## load necessary libraries
```{r}
library(tidyverse)
library(cowplot)
```

## Define plotting functions

```{r}
get_cum_pos <- function(x){
  x_new <- group_by(x, lgN) %>%
    # Compute lg size
    summarise(lg_length=max(pos)) %>%
    # Calculate cumulative position of each chromosome
    mutate(tot=cumsum(lg_length)-lg_length) %>%
    # Add this info to the initial dataset
    left_join(x, ., by=c("lgN")) %>%
    # Add a cumulative position of each SNP
    arrange(lgN, pos) %>%
    group_by(lgN) %>%
    mutate(pos_cum=pos+tot)
  return(x_new)
}

plot_stat_per_snp <- function(x){
  ggplot(x, aes(x=pos, y=untrans_pairwise)) +
    geom_point(color = "black", size=size, alpha=0.5) +
    scale_color_manual(values = c("black","red")) +
    geom_smooth(color='skyblue', se=F) +
    geom_vline(xintercept = c(10744898, 10745093, 10747909, 10749198, 10749404, 10749513, 10749636, 10749663, 10750169, 10767542, 10767560, 10767619, 10767683, 10767684, 10805120, 10841536, 10855605, 10869842, 10872551, 10892267, 10894263, 10948534, 10972255, 11049895, 11050165, 11050174)) +
    scale_x_continuous(breaks=seq(0, 50, 5)) +
    scale_x_continuous(breaks=seq(0, 50, 5)) +
    coord_cartesian(ylim=c(0, 1), expand = F) +
    xlab('position (Mbp)') +
    ylab('Nucleotide Diversity') +
    facet_grid(.~lgN, scales='free_x', space='free_x') +
    theme_cowplot() +
    theme(panel.spacing = unit(0.1, 'lines'),
          axis.title.x=element_text(),
          legend.position='none',
          axis.ticks.x = element_blank(),
          axis.text.x = element_blank(),
          text = element_text(size=13),
          axis.text = element_text(size=10))
}

plot_stat_per_snp_test <- function(x){
  ggplot(x, aes(x=pos, y=correct_fst)) +
    geom_point(color = "black", size=size, alpha=0.5) +
    scale_color_manual(values = c("black","red")) +
    geom_smooth(color='skyblue', se=F) +
    geom_vline(xintercept = c(10744898, 10745093, 10747909, 10749198, 10749404, 10749513, 10749636, 10749663, 10750169, 10767542, 10767560, 10767619, 10767683, 10767684, 10805120, 10841536, 10855605, 10869842, 10872551, 10892267, 10894263, 10948534, 10972255, 11049895, 11050165, 11050174)) +
    scale_x_continuous(breaks=seq(0, 50, 5)) +
    coord_cartesian(ylim=c(0, 1), expand = F) +
    xlab('position (Mbp)') +
    ylab('Fst') +
    facet_grid(.~lgN, scales='free_x', space='free_x') +
    theme_cowplot() +
    theme(panel.spacing = unit(0.1, 'lines'),
          axis.title.x=element_text(),
          legend.position='none',
          axis.text.x = element_blank(),
          text = element_text(size=13),
          axis.text = element_text(size=10))
}
```

## Read in data
```{r}
chr100_div <- read_tsv("/fs/cbsubscb16/storage/rkc/angsd/diversity/PCAM-PPLA_CM023352.1_GOA_polymorphic_folded.thetas.txt") %>% rename(lgN = '#Chromo', pos = Pos) %>% mutate(untrans_pairwise = exp(Pairwise)) %>% filter(pos > 10600000, pos < 11200000)
size = 3

chr100_div_plot <- get_cum_pos(chr100_div) %>% 
  plot_stat_per_snp()
chr100_div_plot
ggsave("/fs/cbsubscb16/storage/rkc/figures/GOA_chr100_div.png", device = "png", width = 7, height = 3)
```

#### Plot fst scan for reference
EastBering-GOA
```{r}
chr100_fst_EastBering_GOA <- read_tsv("/fs/cbsubscb16/storage/rkc/angsd/fst/EastBering-GOA.fst.txt") %>%
  filter(chr == "CM023352.1", midPos > 10600000, midPos < 11200000) %>% rename(lgN = chr, pos = midPos) %>% mutate(correct_fst = ifelse(fst < 0, 0, fst)) %>% select(lgN, pos, correct_fst)

chr100_fst_EastBering_GOA %>% arrange(desc(correct_fst))

chr100_fst_plot <- get_cum_pos(chr100_fst_EastBering_GOA) %>% plot_stat_per_snp_test()
chr100_fst_plot
ggsave("/fs/cbsubscb16/storage/rkc/figures/GOA-EastBering_chr100_fst.png", device = "png", height = 3, width = 7)
```
AI-GOA
```{r}
chr100_fst_AI_GOA <- read_tsv("/fs/cbsubscb16/storage/rkc/angsd/fst/PCAM-PPLA_CM023352.1_AI-GOA_polymorphic_folded.sfs.pbs.fst.txt") %>%
  filter(chr == "CM023352.1", midPos > 10600000, midPos < 11200000) %>% rename(lgN = chr, pos = midPos) %>% mutate(correct_fst = ifelse(fst < 0, 0, fst)) %>% select(lgN, pos, correct_fst)

chr100_fst_AI_GOA %>% arrange(desc(correct_fst))

chr100_fst_plot <- get_cum_pos(chr100_fst_AI_GOA) %>% plot_stat_per_snp_test()
chr100_fst_plot
ggsave("/fs/cbsubscb16/storage/rkc/figures/AI-GOA_chr100_fst.png", device = "png", height = 3, width = 7)
```

GOA-NorthBering
```{r}
chr100_fst_NorthBering_GOA <- read_tsv("/fs/cbsubscb16/storage/rkc/angsd/fst/PCAM-PPLA_CM023352.1_GOA-NorthBering_polymorphic_folded.sfs.pbs.fst.txt") %>%
  filter(chr == "CM023352.1", midPos > 10600000, midPos < 11200000) %>% rename(lgN = chr, pos = midPos) %>% mutate(correct_fst = ifelse(fst < 0, 0, fst)) %>% select(lgN, pos, correct_fst)

chr100_fst_NorthBering_GOA %>% arrange(desc(correct_fst))

chr100_fst_plot <- get_cum_pos(chr100_fst_NorthBering_GOA) %>% plot_stat_per_snp_test()
chr100_fst_plot
ggsave("/fs/cbsubscb16/storage/rkc/figures/GOA-NorthBering_chr100_fst.png", device = "png", height = 3, width = 7)
```

GOA-SEAK
```{r}
chr100_fst_SEAK_GOA <- read_tsv("/fs/cbsubscb16/storage/rkc/angsd/fst/PCAM-PPLA_CM023352.1_GOA-SEAK_polymorphic_folded.sfs.pbs.fst.txt") %>%
  filter(chr == "CM023352.1", midPos > 10600000, midPos < 11200000) %>% rename(lgN = chr, pos = midPos) %>% mutate(correct_fst = ifelse(fst < 0, 0, fst)) %>% select(lgN, pos, correct_fst)

chr100_fst_SEAK_GOA %>% arrange(desc(correct_fst))

chr100_fst_plot <- get_cum_pos(chr100_fst_SEAK_GOA) %>% plot_stat_per_snp_test()
chr100_fst_plot
ggsave("/fs/cbsubscb16/storage/rkc/figures/GOA-SEAK_chr100_fst.png", device = "png", height = 3, width = 7)
```

## Shared loci in top candidate regions across all comparisons
```{r}
list <- c(chr100_fst_AI_GOA,chr100_fst_NorthBering_GOA,chr100_fst_SEAK_GOA)
cand_loci <- rbind(arrange(chr100_fst_AI_GOA, desc(correct_fst))[1:10,],arrange(chr100_fst_EastBering_GOA, desc(correct_fst))[1:10,], arrange(chr100_fst_NorthBering_GOA, desc(correct_fst))[1:10,], arrange(chr100_fst_SEAK_GOA, desc(correct_fst))[1:10,]) %>% group_by(pos) %>% slice(1) %>% select(pos)

#write.table(cand_loci, "/fs/cbsubscb16/storage/rkc/sample_lists/fst_cand_loci.txt")
```

