---
title: "LD-network inversion detection"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load relevant libraries
```{r}
library(LDna)
library(tidyverse)
library(gtools)
library(reshape2)
library(LDheatmap)
source("/fs/cbsubscb16/storage/rkc/scripts/plotLDnetwork_csj.R")
```

## Make LD matrix for chromosome 54 (most likely inversion)

```{bash eval = F, include = T}
zcat angsd/PCAM-PPLA-wholegenome_polymorphic_CM023306.1.beagle.gz | cut -f 1 | sed 's/_/\t/1' | gzip > angsd/global_snp_list_PCAM-PPLA-wholegenome_polymorphic_CM023306.1.txt.gz
```

### Downsample the mafs file and then the beagle file
```{r downsample}
## Read in mafs file
mafs <- read_tsv("/fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA_CM023306.1_global.mafs.gz")
## Total number of SNPs
nrow(mafs)
```

```{r keep every 10 SNPs}
# try to reduce number of SNPs by removing knownEM < 0.01
mafs_filteredEM <- mutate(mafs, 
                        keep = knownEM >= 0.01, 
                        row_number = ifelse(keep, cumsum(keep), -1)) %>% 
  dplyr::select(-row_number)

## number of SNPs left
filter(mafs_filteredEM, keep == T) %>% nrow()
```

```{r}
mafs_filteredEM %>%
  filter(keep==T) %>%
  dplyr::select(1:2) %>%
  write_tsv("/fs/cbsubscb16/storage/rkc/angsd/global_snp_list_PCAM-PPLA_CM023306.1_global_downsampledEM.txt.gz", col_names = F)
(which(mafs_filteredEM$keep)+1) %>% # +1 is necessary because the beagle file has a title line
  as.integer() %>%
  write_lines("/fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA_CM023306.1_global_downsampledEM.pos.idx")
```

select sites from beagle file
```{bash eval = F, include = T}
zcat /fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA-wholegenome_polymorphic_CM023306.1.beagle.gz | cut -f 4- | awk 'NR==FNR{ pos[$1]; next }FNR in pos' /fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA_CM023306.1_global_downsampledEM.pos.idx - | sed 's/_/\t/1' | gzip > /fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA-wholegenome_polymorphic_CM023306.1_downsampledEM.beagle.gz

# make global snplist 
#zcat /fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA-wholegenome_polymorphic_CM023306.1.beagle.gz | awk #'NR==FNR{ pos[$1]; next }FNR in pos' #/fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA_CM023306.1_global_downsampledEM.pos.idx - | cut -f 1 | sed #'s/_/\t/1' | gzip > /fs/cbsubscb16/storage/rkc/angsd/global_snp_list_PCAM-PPLA-wholegenome_polymorphic#_CM023306.1_downsampledEM.txt.gz

# remove first line
#zcat angsd/PCAM-PPLA-wholegenome_polymorphic_CM023306.1_downsampledEM.beagle.gz | tail -n +2 | gzip > #angsd/PCAM-PPLA-wholegenome_polymorphic_CM023306.1_downsampledEM_noheader.beagle.gz 
```

#### Run ngsLD 
```{bash eval = F, include = T}
## Full snp list
## important updates to script:
# Path to ngsLD has changed to '/programs/ngsLD-1.1.1/ngsld.sif /ngsLD-1.1.1/ngsLD'. On top of this, it has been installed in something called a 'Singularity' which requires us to specify a working directory for the program that is done using 'export APPTAINER_BINDPATH=/fs/cbsubscb16/storage/rkc/'

export APPTAINER_BINDPATH=/fs/cbsubscb16/storage/rkc/

nohup bash /fs/cbsubscb16/storage/genomic-data-analysis/scripts/get_ld.sh /fs/cbsubscb16/storage/rkc/angsd/ PCAM-PPLA-wholegenome_polymorphic_CM023306.1_noheader.beagle.gz global_snp_list_PCAM-PPLA-wholegenome_polymorphic_CM023306.1.txt.gz 125 16 '/programs/ngsLD-1.1.1/ngsld.sif /ngsLD-1.1.1/ngsLD' > /fs/cbsubscb16/storage/rkc/nohups/run_ngsLD_bam_list_PCAM-PPLA-wholegenome_polymorphic_CM023306.1_125.nohup &
```

```{bash eval = F, include = T}
## downsampled snp list
## important updates to script:
# Path to ngsLD has changed to '/programs/ngsLD-1.1.1/ngsld.sif /ngsLD-1.1.1/ngsLD'. On top of this, it has been installed in something called a 'Singularity' which requires us to specify a working directory for the program that is done using 'export APPTAINER_BINDPATH=/fs/cbsubscb16/storage/rkc/'

export APPTAINER_BINDPATH=/fs/cbsubscb16/storage/rkc/

nohup bash /fs/cbsubscb16/storage/genomic-data-analysis/scripts/get_ld.sh /fs/cbsubscb16/storage/rkc/angsd/ PCAM-PPLA-wholegenome_polymorphic_CM023306.1_downsampledEM_noheader.beagle.gz global_snp_list_PCAM-PPLA-wholegenome_polymorphic_CM023306.1_downsampledEM.txt.gz 15000 16 '/programs/ngsLD-1.1.1/ngsld.sif /ngsLD-1.1.1/ngsLD' > /fs/cbsubscb16/storage/rkc/nohups/run_ngsLD_bam_list_PCAM-PPLA-wholegenome_polymorphic_CM023306.1_downsampledEM_15000.nohup &
```

## Plot LD heatmap to see if there is an obvious block of ld
Just plotting the start because there is too much data. Should downsample later
- window should lie between positions 19841923-24981505 on chr 54

```{bash eval=F, include=T}
cat /fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA-wholegenome_polymorphic_CM023306.ld | bash /fs/cbsubscb16/storage/rkc/scripts/ld_blocks.sh \
CM023306.1 \
19000000 \
36000000
```

Do not see the start of an ld block here. 

## LDna

#### make matrix from ngsLD output

```{bash eval=F, include=T}
# Extract sites from input file
cat /fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA-wholegenome_polymorphic_CM023306.ld | awk -v chr=CM023306.1 -v min=19000000 -v max=36000000 'BEGIN{print "snp1\tsnp2\tdist\tr2p\tD\tDp\tr2"} {split($1,pos1,":"); split($2,pos2,":")} pos1[1]==chr && pos1[2]>=min && pos1[2]<=max && pos2[1]==chr && pos2[2]>=min && pos2[2]<=max' | cut -f 1-7 > /fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA_CM023306.1_downsampledEM_15000_premat.ld
```

```{r}
r <- read.table("/fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA_CM023306.1_downsampledEM_15000_premat.ld", header=TRUE, stringsAsFactors=FALSE)
id <- unique(mixedsort(c(r[,"snp1"],r[,"snp2"])))
posStart <- head(id,1)
posEnd <- tail(id,1)
r <- rbind(r, c(posStart,posStart,0,NA,NA,NA,NA), c(posEnd,posEnd,0,NA,NA,NA,NA))

for (ld in c("r2")){
  m <- apply(acast(r, snp1 ~ snp2, value.var=ld, drop=FALSE),2,as.numeric)
  rownames(m) <- colnames(m)
  m <- m[mixedorder(rownames(m)),mixedorder(colnames(m))]
  id <- rownames(m)
  dist <- as.numeric(sub(".*:","",id))
}

mode(m) = "numeric"
# m_df <- as.data.frame(m)
# write_tsv(m_df, "/fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA-wholegenome_polymorphic_downsampledEM.mat")
```

## Run LDna
```{r}
ldna <- LDnaRaw(m)
LD_clusters <- as.data.frame(ldna$clusterfile)
ldna$stats
```

#### Extract clusters
```{r}
phi <- 6
par(mfcol=c(1,2))
cluster1 <- extractClusters(ldna, m,min.edges = 5, phi = 6)
```

```{r}
#summarize LDna analysis
summary1 <- summaryLDna(ldna, cluster1, m)
summary1
```
```{r}
par(mfcol=c(1,2))
plotLDnetwork_csj(ldna, LDmat=m, option = 2, clusters = cluster1, summary = summary1)
```

```{r}
par(mfcol=c(1,3))
extractClusters(ldna, m, min.edges = 1, plot.tree = T, extract = F)
extractClusters(ldna, m, min.edges = 3, plot.tree = T, extract = F)
extractClusters(ldna, m, min.edges = 5, plot.tree = T, extract = F)
```
Looks like there are two obvious clusters from the above plot. Lets try playing with thresholds around 0.35

```{r}
par(mfcol=c(1,2))
plotLDnetwork_csj(LDmat=m, option = 1, threshold = 0.37)
plotLDnetwork_csj(LDmat=m, option = 1, threshold = 0.34)
```
#### Color sites by cluster assignment
```{r}
LD_clusters <- as.data.frame(ldna$clusterfile)
LD_clusters <- LD_clusters %>% mutate(site = rownames(LD_clusters), .before = `182_0.11`) %>% select(site, `138_0.38`, `139_0.38`) %>% separate(site, into = c("chr", "pos"), sep = ":") %>% 
  mutate(cluster = ifelse(`138_0.38` == T, "138_0.38", ifelse(`139_0.38` == T, "139_0.38", NA)))
LD_clusters$pos <- as.numeric(LD_clusters$pos)

LD_clusters %>% ggplot(aes(x = pos, y = chr)) + 
  geom_point(aes(color = cluster), pch = "+", size = 5 )  +
  scale_x_continuous(breaks = seq(19500000,31960000, 2000000)) +
  geom_vline(xintercept = c(20520440, 29903464), color = "black", alpha = 0.5)

ggsave("/fs/cbsubscb16/storage/rkc/figures/ld_window_clusters.png", device = "png", height = 4, width = 7, units = "in")
```
