 ---
title: "Local PCA"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load required packages
```{r}
library(tidyverse)
```

## Subset beagle into chromosomes

#### Make lg list

```{bash eval=T, include=F}
zcat angsd/PCAM-PPLA-wholegenome_polymorphic.beagle.gz | sed 's/\_.*//' | uniq > sample_lists/lg_list.txt
```

#### Run subsetting script

```{bash index bam files, eval=F, include=T}
## Run on cluster
## Job ID 48750

echo '#!/bin/bash 
bash /fs/cbsubscb16/storage/genomic-data-analysis/scripts/subset_beagle_by_lg.sh /fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA-wholegenome_polymorphic.beagle.gz /fs/cbsubscb16/storage/rkc/sample_lists/lg_list.txt 24' | \
sbatch \
--nodelist=cbsubscb16 \
--partition=regular \
--nodes=1 \
--ntasks=25 \
--mem=5G \
--output=/home/cas399/rkc/log/subset_beagle.log
```

## Run local PCA 10000 SNP window

```{bash eval=T, include=F}
## Run on cluster
## current issue, new pcangsd flags differ. will have to modify local_pca_1.sh script. 
## Job ID 49957

echo '#!/bin/bash 
bash /fs/cbsubscb16/storage/genomic-data-analysis/scripts/run_local_pca.sh /fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA-wholegenome_polymorphic.beagle.gz /fs/cbsubscb16/storage/rkc/sample_lists/lg_list.txt 10000 2 26 /fs/cbsubscb16/storage/genomic-data-analysis/scripts/local_pca_1.sh /fs/cbsubscb16/storage/genomic-data-analysis/scripts/local_pca_2.R python /programs/pcangsd-1.10/pcangsd/pcangsd.py' | \
sbatch \
--nodelist=cbsubscb16 \
--partition=regular \
--nodes=1 \
--ntasks=26 \
--mem=5G \
--output=/home/cas399/rkc/log/local_pca_all_chr.log
```

### Assemble local_pca input

```{r warning=F, message=F}
lg_list <- read_lines("/fs/cbsubscb16/storage/rkc/sample_lists/lg_list.txt")
i=1
for (lg in lg_list){
  pca_summary_temp <- read_tsv(paste0("/fs/cbsubscb16/storage/rkc/angsd/local_pca/pca_summary_10000snp_", lg, "_2pc.tsv"), col_names = F)
  snp_position_temp <- read_tsv(paste0("/fs/cbsubscb16/storage/rkc/angsd/local_pca/snp_position_10000snp_", lg, "_2pc.tsv"), col_names = F)
  if (i == 1) {
    pca_summary <- pca_summary_temp
    snp_position <- snp_position_temp
  } else {
    pca_summary <- bind_rows(pca_summary, pca_summary_temp)
    snp_position <- bind_rows(snp_position, snp_position_temp)
  }
  i <- i+1
}
#write_tsv(pca_summary, "/fs/cbsubscb16/storage/rkc/angsd/local_pca/pca_summary_10000snp_2pc.tsv", col_names = F)
#write_tsv(snp_position, "/fs/cbsubscb16/storage/rkc/angsd/local_pca/snp_position_10000snp_2pc.tsv", col_names = F)
```

### Run pc_dist and assemble the output

```{r message=F, eval=F, echo=T}
rscript <- "library(tidyverse)
library(lostruct)
## Read the input
pca_summary <- read_tsv('/fs/cbsubscb16/storage/rkc/angsd/local_pca/pca_summary_10000snp_2pc.tsv', col_names = F) 
## Run pc_dist with pca_summary
pca_summary <- as.matrix(pca_summary)
attr(pca_summary, 'npc') <- 2
dist <- pc_dist(pca_summary)
write_tsv(as.data.frame(dist), '/fs/cbsubscb16/storage/rkc/angsd/local_pca/window_dist_10000snp_2pc.tsv', col_names = F)"
write_lines(rscript, "/fs/cbsubscb16/storage/rkc/scripts/run_pc_dist_10000snp_2pc.R")
```

```{r message=F, eval=F, echo=T}
nohup Rscript /fs/cbsubscb16/storage/rkc/scripts/run_pc_dist_10000snp_2pc.R \
> /fs/cbsubscb16/storage/rkc/nohups/run_pc_dist_10000snp_2pc.nohup 2>&1 &
```

### Analysis with all windows

```{r fig.height=9, fig.width=12, message=F, warning=F}
## Read the SNP position file
snp_position <- read_tsv("/fs/cbsubscb16/storage/rkc/angsd/local_pca/snp_position_10000snp_2pc.tsv", col_names = F) %>%
  transmute(lg = sub("\\_.*", "", X1), 
            start = as.integer(sub(".*\\_", "", X1)), 
            stop = as.integer(sub(".*\\_", "", X2)), 
            center = (start+stop)/2)
## Read in the window distance matrix
dist <- read_tsv("/fs/cbsubscb16/storage/rkc/angsd/local_pca/window_dist_10000snp_2pc.tsv", col_names = F) %>%
  as.matrix()
## Do MDS with the distance matrix
mds <- cmdscale(as.dist(dist), k=10, eig = T)
proportion_variance <- round(mds$eig/sum(mds$eig)*100,2)
mds <- mds$points %>%
  as.data.frame() %>%
  set_names(paste0("dist_", 1:10)) %>%
  bind_cols(snp_position, .)
mds %>%
  gather(key = axis, value = value, 5:14) %>%
  mutate(axis=fct_relevel(axis, paste0("dist_", 1:10))) %>%
  ggplot(aes(x=center/10^6, y=value, color=axis)) +
  geom_point(size=0.5) +
  facet_grid(axis~lg, scales="free_x", space="free_x") +
  theme_bw() +
  xlab("position (Mbp)") +
  ylab("principal coordinate value") +
  theme(panel.spacing = unit(0.1, "lines"),
      axis.title.x=element_text(),
      legend.position="none",
      text = element_text(size=10),
      axis.text = element_text(size=6),
      panel.grid.major = element_blank(), 
      panel.grid.minor = element_blank()) 

ggsave("/fs/cbsubscb16/storage/rkc/figures/local_pca_scan.png", plot = last_plot(), device = "png", width = 20, height = 10, units = "in")
ggplot(mds, aes(x=dist_1, y=dist_2, color=lg, shape=lg)) +
  geom_point() +
  scale_shape_manual(values = c(rep(c(15,16,17,18),26))) +
  xlab(paste0("PCo1 (", proportion_variance[1], "%)")) +
  ylab(paste0("PCo2 (", proportion_variance[2], "%)")) +
  cowplot::theme_cowplot()
ggplot(mds, aes(x=dist_3, y=dist_4, color=lg, shape=lg)) +
  geom_point() +
  scale_shape_manual(values = c(rep(c(15,16,17,18),26))) +
  xlab(paste0("PCo3 (", proportion_variance[3], "%)")) +
  ylab(paste0("PCo4 (", proportion_variance[4], "%)")) +
  cowplot::theme_cowplot()
ggplot(mds, aes(x=dist_5, y=dist_6, color=lg, shape=lg)) +
  geom_point() +
  scale_shape_manual(values = c(rep(c(15,16,17,18),26))) +
  xlab(paste0("PCo5 (", proportion_variance[5], "%)")) +
  ylab(paste0("PCo6 (", proportion_variance[6], "%)")) +
  cowplot::theme_cowplot()
ggplot(mds, aes(x=dist_7, y=dist_8, color=lg, shape=lg)) +
  geom_point() +
  scale_shape_manual(values = c(rep(c(15,16,17,18),26))) +
  xlab(paste0("PCo7 (", proportion_variance[7], "%)")) +
  ylab(paste0("PCo8 (", proportion_variance[8], "%)")) +
  cowplot::theme_cowplot()
```

## Run local PCA 1000 SNP window

```{bash eval=T, include=F}
## Run on cluster
## Job ID 

echo '#!/bin/bash 
bash /fs/cbsubscb16/storage/genomic-data-analysis/scripts/run_local_pca.sh /fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA-wholegenome_polymorphic.beagle.gz /fs/cbsubscb16/storage/rkc/sample_lists/lg_list.txt 1000 2 26 /fs/cbsubscb16/storage/genomic-data-analysis/scripts/local_pca_1.sh /fs/cbsubscb16/storage/genomic-data-analysis/scripts/local_pca_2.R python /programs/pcangsd-1.10/pcangsd/pcangsd.py' | \
sbatch \
--nodelist=cbsubscb16 \
--partition=regular \
--nodes=1 \
--ntasks=26 \
--mem=5G \
--output=/home/cas399/rkc/log/local_pca_all_chr_1000.log
```
