---
title: "SNP Heatmap"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## load appropriate libraries and data 
```{r message=F, warning=F}
library(tidyverse)
library(knitr)
sample_table <- read_tsv("/fs/cbsubscb16/storage/rkc/sample_lists/sample_table.tsv", col_types = c(rep("c",4),"d","d",rep("c",3)))

## Define a function to get allele dosage from genotype likelihood
get_allele_dosage <- function(x){
  if(x[2]==x[3] & x[3]==x[4]){
    return(NA) ## return NA at sites with no coverages (where rescaled GLs are 0.333, 0.333, 0.333)
  } else {
    if(x[1]>0.5) { ## polarize the GLs based on frequency at reference individual
      x[2:4] <- x[4:2]
    }
    return(x[3] + 2*x[4])
  }
}
```

```{bash eval=F, include=T}
## subset outlier region on chromosome 100
## outlier region is between 10744898-11050174 which corresponds with lines 14821-15070.
## We'll capture adjacent regions too: 10600000-11200000, corresponding to 
zcat angsd/PCAM-PPLA-wholegenome_polymorphic_CM023352.1.beagle.gz | sed -n -e '1p' -e '14635,15268p;15269q' | gzip > angsd/PCAM-PPLA-wholegenome_polymorphic_CM023352.1_outlierPlus.beagle.gz
```

```{r message=F}
# read beagle file
outlier_beagle <- read_tsv("/fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA-wholegenome_polymorphic_CM023352.1_outlierPlus.beagle.gz") #%>% dplyr::select(-c(Ind336...505,Ind336...506,Ind336...507))
```

## Polarize genotype probabilities
```{r}
# We will polarize by a Gulf of Alaska population with no admixture with East Bering. Let's use ABLG5592
# reduce outlier beagle to first 30 row to test script
outlier_beagle_30 <- head(outlier_beagle, 30)

```

```{r}
# try nicolas' and liams script for extracting allelic dosage and polarize in same step
# separate linkage group and position 
outlier_beagle <- outlier_beagle %>% separate(marker, c(NA ,"lg", "pos")) %>% 
  dplyr::select(-allele2)

# set up the data frame, ncols should equal the number of samples and nrows should equal the number of loci
dosage <- as.data.frame(matrix(ncol = (ncol(outlier_beagle)-3)/3+3, nrow = nrow(outlier_beagle)))
dosage[,1:3]=outlier_beagle[,1:3]

## Get allele dosage iteratively from each individual
# 4 columns are extracted for each individual. the 1st of the four columns is the GL for the major allele for the polarizing individual. Choose the reference column according to the individual you want to polarize from. In our case, that is column 67 (ABLG5418)
for (j in 1:((ncol(outlier_beagle)-3)/3)){
  temp_ind <- outlier_beagle[,c(67, (1+3*j):(3+3*j))]
  dosage[,3+j] <- apply(temp_ind, 1, get_allele_dosage)
}

# rename columns to match sample names

colnames(dosage) <- c("V1","V2","V3",as.double(sample_table$ABLG))
```

#### Convert to long format for plotting

```{r}

dosage_long <- pivot_longer(dosage, cols = 4:ncol(dosage), names_to = "ABLG", values_to = "allele_dosage") %>% 
  dplyr::select(-V3) %>% 
  rename(lg = V1, pos = V2) %>% 
  left_join(dplyr::select(sample_table, population, Loc, ABLG), by = "ABLG")

#write_tsv(dosage_long, "/fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA-wholegenome_polymorphic_CM023352.1_outlierPlus.alleledosage.tsv")
```

## Plot genotype heatmaps

```{r}
#dosage_long <- read_tsv("/fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA-wholegenome_polymorphic_CM023352.1_outlierPlus.alleledosage.tsv")

dosage_long_plot <- dosage_long %>% 
  mutate(allele_dosage_rounded = as.character(round(allele_dosage))) %>% 
  ggplot(aes(x=as_factor(pos), y=ABLG, fill = allele_dosage_rounded)) +
  geom_tile() +
  scale_fill_viridis_d(direction = 1) +
  facet_grid(Loc~lg, scales = "free", space = "free") +
  theme_bw() +
  theme(axis.text = element_blank(),
        axis.title = element_blank(),
        axis.ticks.x = element_blank(),
        axis.ticks.y = element_blank(),
        panel.spacing.x = unit(0.2, "lines"),
        strip.text.y = element_text(angle = 0))
dosage_long_plot

#ggsave("/fs/cbsubscb16/storage/rkc/figures/snp_heatmap_chr100_outlier.png", device = "png",width = 11, height = 8.5)
```

## Zoom in on outlier area only

```{bash eval=F, include=T}
## subset outlier region on chromosome 100
## outlier region is between 10744898-11050174 which corresponds with lines 14821-15070.
zcat angsd/PCAM-PPLA-wholegenome_polymorphic_CM023352.1.beagle.gz | sed -n -e '1p' -e '14821,15070p;15071q' | gzip > angsd/PCAM-PPLA-wholegenome_polymorphic_CM023352.1_outlierZoom.beagle.gz
```

```{r message=F}
# read beagle file
outlier_beagle_zoom <- read_tsv("/fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA-wholegenome_polymorphic_CM023352.1_outlierZoom.beagle.gz") 
```

## Polarize genotype probabilities

```{r}
# try nicolas' and liams script for extracting allelic dosage and polarize in same step
# separate linkage group and position 
outlier_beagle_zoom <- outlier_beagle_zoom %>% separate(marker, c(NA ,"lg", "pos")) %>% 
  dplyr::select(-allele2)

# set up the data frame, ncols should equal the number of samples and nrows should equal the number of loci
dosage_zoom <- as.data.frame(matrix(ncol = (ncol(outlier_beagle_zoom)-3)/3+3, nrow = nrow(outlier_beagle_zoom)))
dosage_zoom[,1:3]=outlier_beagle_zoom[,1:3]

## Get allele dosage iteratively from each individual
# 4 columns are extracted for each individual. the 1st of the four columns is the GL for the major allele for the polarizing individual. Choose the reference column according to the individual you want to polarize from. In our case, that is column 67 (ABLG5418)
for (j in 1:((ncol(outlier_beagle_zoom)-3)/3)){
  temp_ind <- outlier_beagle_zoom[,c(67, (1+3*j):(3+3*j))]
  dosage_zoom[,3+j] <- apply(temp_ind, 1, get_allele_dosage)
}

# rename columns to match sample names

colnames(dosage_zoom) <- c("V1","V2","V3",as.double(sample_table$ABLG))
```

#### Convert to long format for plotting

```{r}
dosage_zoom_long <- pivot_longer(dosage_zoom, cols = 4:ncol(dosage_zoom), names_to = "ABLG", values_to = "allele_dosage") %>% 
  dplyr::select(-V3) %>% 
  rename(lg = V1, pos = V2) %>% 
  left_join(dplyr::select(sample_table, population, Loc, ABLG), by = "ABLG")

#write_tsv(dosage_zoom_long, "/fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA-wholegenome_polymorphic_CM023352.1_outlierZoom.alleledosage.tsv")
```

## Plot genotype heatmaps

```{r}
#dosage_zoom_long <- read_tsv("/fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA-wholegenome_polymorphic_CM023352.1_outlierZoom.alleledosage.tsv")

dosage_zoom_long_plot <- dosage_zoom_long %>% 
  mutate(allele_dosage_rounded = as.character(round(allele_dosage))) %>% 
  ggplot(aes(x=as_factor(pos), y=ABLG, fill = allele_dosage_rounded)) +
  geom_tile() +
  scale_fill_manual(values = c("gold","darkorange3","firebrick4"), na.value = "grey80") +
  facet_grid(Loc~lg, scales = "free", space = "free") +
  theme_bw() +
  theme(axis.text = element_blank(),
        axis.title = element_blank(),
        axis.ticks.x = element_blank(),
        axis.ticks.y = element_blank(),
        panel.spacing.x = unit(0.2, "lines"),
        strip.text.y = element_text(angle = 0))
dosage_zoom_long_plot

#ggsave("/fs/cbsubscb16/storage/rkc/figures/snp_heatmap_chr100_outlierZoom.png", device = "png",width = 11, height = 8.5)
```

## Just produce heatmap for 10 highest fst candidate loci from each pairwise comparison

```{r message=F}
# read beagle file
outlier_beagle_cand <- read_tsv("/fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA-wholegenome_polymorphic_CM023352.1_outlierCandidates.beagle.gz") 
```

```{r}
# try nicolas' and liams script for extracting allelic dosage and polarize in same step
# separate linkage group and position 
outlier_beagle_cand <- outlier_beagle_cand %>% separate(marker, c(NA ,"lg", "pos")) %>% 
  dplyr::select(-allele2) %>% mutate(lg = "49")

# set up the data frame, ncols should equal the number of samples and nrows should equal the number of loci
dosage_cand <- as.data.frame(matrix(ncol = (ncol(outlier_beagle_cand)-3)/3+3, nrow = nrow(outlier_beagle_cand)))
dosage_cand[,1:3]=outlier_beagle_cand[,1:3]

## Get allele dosage iteratively from each individual
# 4 columns are extracted for each individual. the 1st of the four columns is the GL for the major allele for the polarizing individual. Choose the reference column according to the individual you want to polarize from. In our case, that is column 67 (ABLG5418)
for (j in 1:((ncol(outlier_beagle_cand)-3)/3)){
  temp_ind <- outlier_beagle_cand[,c(67, (1+3*j):(3+3*j))]
  dosage_cand[,3+j] <- apply(temp_ind, 1, get_allele_dosage)
}

# rename columns to match sample names

colnames(dosage_cand) <- c("V1","V2","V3",as.double(sample_table$ABLG))
```

#### Convert to long format for plotting

```{r}
dosage_cand_long <- pivot_longer(dosage_cand, cols = 4:ncol(dosage_cand), names_to = "ABLG", values_to = "allele_dosage") %>% 
  dplyr::select(-V3) %>% 
  rename(lg = V1, pos = V2) %>% 
  left_join(dplyr::select(sample_table, population, Loc, ABLG), by = "ABLG")

#write_tsv(dosage_zoom_long, "/fs/cbsubscb16/storage/rkc/angsd/PCAM-PPLA-wholegenome_polymorphic_CM023352.1_outlierZoom.alleledosage.tsv")
```

## Plot genotype heatmaps

```{r}
dosage_cand_long_plot <- dosage_cand_long %>% 
  mutate(allele_dosage_rounded = as.character(round(allele_dosage))) %>% 
  ggplot(aes(x=as_factor(pos), y=ABLG, fill = allele_dosage_rounded)) +
  geom_tile() +
  scale_fill_manual(values = c("gold","darkorange3","firebrick4"), na.value = "grey80") +
  facet_grid(Loc~lg, scales = "free", space = "free") +
  theme_bw() +
  theme(axis.text.y = element_blank(),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
        axis.title = element_blank(),
        axis.ticks.x = element_blank(),
        axis.ticks.y = element_blank(),
        panel.spacing.x = unit(0.2, "lines"),
        strip.text.y = element_text(angle = 0))
dosage_cand_long_plot

# ggsave("/fs/cbsubscb16/storage/rkc/figures/snp_heatmap_chr100_outlierCandidates.png", device = "png",width = 11, height = 8.5)
```

## Plot genotype heatmaps continuous

```{r}
dosage_cand_long_plot <- dosage_cand_long %>% 
  ggplot(aes(x=as_factor(pos), y=ABLG, fill = allele_dosage)) +
  geom_tile() +
  scale_fill_distiller(type = "seq", palette = 18) +
  facet_grid(Loc~lg, scales = "free", space = "free") +
  theme_bw() +
  theme(axis.text.y = element_blank(),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
        axis.title = element_blank(),
        axis.ticks.x = element_blank(),
        axis.ticks.y = element_blank(),
        panel.spacing.x = unit(0.2, "lines"),
        strip.text.y = element_text(angle = 0))
dosage_cand_long_plot

ggsave("/fs/cbsubscb16/storage/rkc/figures/snp_heatmap_chr100_outlierCandidates.png", device = "png",width = 11, height = 8.5)
```